MatrixSDK/Background/MXBackgroundSyncService.swift

```diff
@@ -0,0 +1,521 @@
+            decryptionResult.senderCurve25519Key = senderKey
+            decryptionResult.claimedEd25519Key = claimedKeys?["ed25519"] as? String
+            event.setClearData(decryptionResult)
+        } else {
+            throw MXBackgroundSyncServiceError.unknownAlgorithm
+        }
+    }
+    
+    private func decryptMessageWithOlm(message: [AnyHashable: Any], theirDeviceIdentityKey: String) -> String? {
+        let sessionIds = olmDevice.sessionIds(forDevice: theirDeviceIdentityKey)
+        let messageBody = message["body"] as? String
+        let messageType = message["type"] as? UInt ?? 0
+        
+        for sessionId in sessionIds ?? [] {
+            if let payload = olmDevice.decryptMessage(messageBody,
+                                                      withType: messageType,
+                                                      sessionId: sessionId,
+                                                      theirDeviceIdentityKey: theirDeviceIdentityKey) {
+                return payload
+            } else {
+                let foundSession = olmDevice.matchesSession(theirDeviceIdentityKey,
+                                                            sessionId: sessionId,
+                                                            messageType: messageType,
+                                                            ciphertext: messageBody)
+                if foundSession {
+                    return nil
+                }
+            }
+        }
+        
+        if messageType != 0 {
+            return nil
+        }
+        
+        var payload: NSString?
+        guard let _ = olmDevice.createInboundSession(theirDeviceIdentityKey,
+                                                     messageType: messageType,
+                                                     cipherText: messageBody,
+                                                     payload: &payload) else {
+                                                        return nil
+        }
+        return payload as String?
+    }
+    
+    private func handleSyncResponse(_ syncResponse: MXSyncResponse) {
+        NSLog("[MXBackgroundSyncService] handleSyncResponse: Received %tu joined rooms, %tu invited rooms, %tu left rooms, %tu toDevice events.",
+              syncResponse.rooms.join.count,
+              syncResponse.rooms.invite.count,
+              syncResponse.rooms.leave.count,
+              syncResponse.toDevice.events?.count ?? 0)
+        
+        self.pushRulesManager.handleAccountData(syncResponse.accountData)
+        self.updateStore(with: syncResponse)
+        
+        for event in syncResponse.toDevice?.events ?? [] {
+            handleToDeviceEvent(event)
+        }
+        
+        //  update event stream token
+        NSLog("[MXBackgroundSyncService] handleSyncResponse: Next sync token: \(syncResponse.nextBatch ?? "nil")")
+        self.store.eventStreamToken = syncResponse.nextBatch
```

Comment:
Avoid to write to MXStore as explained in another comment.

Reply:
Stopped updating `store.eventStreamToken`.